tests/storage/test_client_ips.py

```diff
@@ -641,14 +644,14 @@ def test_old_user_ips_pruned(self) -> None:
 
         r = result2[(user_id, device_id)]
         self.assertLessEqual(
```

Comment:
This looks suspicious to me. It used to say `assertDictContainsSubset`: https://github.com/matrix-org/synapse/blob/84f441f88f51d3f94e1616e1e5507df0dadb6de8/tests/storage/test_client_ips.py#L643 but this changed in #16099.

Time to dive down the rabbit hole.

# Previous behaviour

At first glance, I thought this used to compare two dictionaries with `<=`. I didn't know you could do this, and apparently you can't:

> Dictionaries compare equal if and only if they have the same (key, value) pairs (regardless of ordering). Order comparisons (‘<’, ‘<=’, ‘>=’, ‘>’) raise [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError).

(https://docs.python.org/3/library/stdtypes.html#dict, under `d |= other`.) On second reading, I saw that this compares the `.items()` with <=. I didn't know you could do this either! In Python3 `.items()` returns [a dict_items view](https://docs.python.org/3/library/stdtypes.html#dict). Apparently

>  If all values are hashable, so that (key, value) pairs are unique and hashable, then the items view is also set-like. [...] For set-like views, all of the operations defined for the abstract base class [collections.abc.Set](https://docs.python.org/3/library/collections.abc.html#collections.abc.Set) are available (for example, ==, <, or ^).

The collections.abc docs say that Set defines a __le__ method. It doesn't define semantics explicitly, but https://docs.python.org/3/library/stdtypes.html?#frozenset.issubset says that this is a subset test. So I think `.items() <= .items()` views both dictionaries as a set of (key, value) pairs and asks if the left is a subset of the right.

# New behaviour

I didn't know you could compare attrs instances with <= either. But apparently you can:

```python
>>> @attr.s(slots=True, frozen=True, auto_attribs=True)
... class C:
...     x: int
...     y: str
... 
>>> C(1, "a") <= C(2, "b")
True
>>> C(1, "a") <= C(0, "")
False
```

:warning:  This does not look like subset semantics! Instead, I'm guessing that `a <= b`  is a lexicographic sort ordering, i.e. a <= b if
 - a.field1 <= b.field1; or
 - a.field1 == b.field1 and a.field2 <= b.field2; or
 - a.field1 == b.field1 and a.field2 == b.field2 and a.field3 <= b.field3; or;
 - ... etc

Attrs docs on comparisons are here: https://www.attrs.org/en/stable/comparison.html. The order comparison operators are defined if pass order=True to attr.s(), and according to the footnote

> That’s the default if you use the [attr.s()](https://www.attrs.org/en/stable/api-attr.html#attr.s) decorator, but not with [define()](https://www.attrs.org/en/stable/api.html#attrs.define).

Unfortunately the attrs docs don't define the semantics of the ordering comparison operators. It only says:

> attrs will also create a full set of ordering methods that are based on the defined fields: __le__, __lt__, __ge__, and __gt__.

Reply:
I'm not fully sure why the original test was doing a subset check. Is there any reason why this can't be a direct `==` comparison of attrs structs?

Reply:
I failed to notice that these were exactly equal checks, it looks unnecessary (maybe it was just being cautions because dicts?)

So yeah, I think these can all just be direct equals.